[llvm-mca][AArch64] Refactor Neoverse tests to split out common inputs (NFC)

c-rhodes opened the pull request:
For many of these tests the inputs are very similiar with slight divergences, ideally we'd have common sources where we can to avoid subtle differences.

Changes:

- For basic tests there's a relatively large diff between V1 and all other cores because of 24f0901 (#128892) which makes it more complete. I've been thru the entire diff and 99% of the time V1 makes more sense, except for a couple of small changes (might post a separate patch for). Therefore I decided it's best to take V1-basic-instructions.s as the common source.
- Split out FEAT_RCPC_IMMO tests from basic since N1 doesnt have this feature.
- Split out FEAT_MTE tests. V2/V3 also have this feature but were missing tests, so I've added them.

The NEON and SVE tests are also quite substantial and could be made common, but there's also subtle differences that take time to go thru.

c-rhodes committed:
[llvm-mca][AArch64] Refactor Neoverse tests to split out common inputs (NFC)

For many of these tests the inputs are very similiar with slight
divergences, we should work to make common sources where we can to avoid
subtle differences.

Changes:

- For basic tests there's a relatively large diff between V1 and all
  other cores because of 24f0901 (#128892) which makes it more complete.
  I've been thru the entire diff and 99% of the time V1 makes more
  sense, except for a couple of small changes (might post a separate
  patch for). Therefore I decided it's best to take
  V1-basic-instructions.s as the common source.
- Split out FEAT_RCPC_IMMO tests from basic since N1 doesnt have this feature.
- Split out FEAT_MTE tests. V2/V3 also have this feature but were
  missing tests, so I've added them.

The NEON and SVE tests are also quite substantial and could be made
common, but there's also subtle differences that take time to go thru.

Asher8118 reviewed (commented):
 I think the change makes sense and I agree that a common source for testing is a good direction.
 
> I've been thru the entire diff and 99% of the time V1 makes more sense, except for a couple of small changes (might post a separate patch for).

One question, in the cases where we do have even slight divergence, what is the plan? Add those instructions specifically as a separate test to avoid duplication?

>The NEON and SVE tests are also quite substantial and could be made common, but there's also subtle differences that take time to go thru

I think as long as exceptions are handled as well, that would still be an improvement to the amount of duplication we currently have in tests.

Asher8118 reviewed (approved):
>>If the divergence is necessary then yes, but I imagine a lot of them arent. For the basic tests there were 3 types of divergence:
>For add(s)/sub(s)/add/cmn/cmp v1 would test shift <= 4 whereas v2/v3 didnt, e.g. add x17, x29, x20, lsl #3 (v1) vs add x17, x29, x20, lsl #63 (v2/v3). Looking at the SWOGs shift <= 4 is a special-case so I think V1 is better.
V2/V3 had quite a lot of duplicate instructions
Better opcode coverage in V1 than other cores.

Thank you for your answer. LGTM

rj-jesus reviewed (commented):
Thanks for doing this - I just have a couple of questions/suggestions. :)

c-rhodes committed:
address comments

- take 16-bit load/store register offset from v2/v3 (better coverage)
- update rcpc-immo header comment

rj-jesus reviewed (approved):
LGTM, cheers!

cofibrant reviewed (commented):
I had accidentally commented on the commit, not the PR, so reposting this comment. Sorry if you've already seen it!
